Add config flow for ping

jpbede opened the pull request:
## Proposed change
Give some more love to the ping integration. Added configuration flow and migration for `binary_sensor` and `device_tracker` platforms. Since both platforms do almost the same thing, I simplified both platforms and used one coordinator.

When migrating, a config entry per host will be created and HA will always set up both platforms. Depending on where the import came from, one of the platforms will be disabled by default.

For newly created entries, the device_tracker platform is disabled by default.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28394

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

GrumpyMeow commented on homeassistant/components/ping/config_flow.py:
```diff
@@ -0,0 +1,99 @@
+    return _set_platform
+
+
+BINARY_SENSOR_OPTIONS_SCHEMA = vol.Schema(
```
It would be great if the update_interval/polling_interval can also be configured via the UI.

joostlek replied:
We don't want that to be configured via UI. If you want to change the amount of times an integration is updated, you can create an automation afaik

jpbede committed:
Add config flow for ping binary sensor

jpbede commented on homeassistant/components/ping/strings.json:
```diff
@@ -4,5 +4,36 @@
+  "config": {
+    "step": {
+      "user": {
+        "title": "Add Ping",
+        "description": "Ping allows you to check the availability of a host.",
+        "menu_options": {
+          "binary_sensor": "Binary sensor",
+          "device_tracker": "Device tracker"
+        }
+      },
+      "binary_sensor": {
+        "title": "[%key:component::ping::config::step::user::menu_options::binary_sensor%]",
+        "data": {
+          "name": "[%key:common::config_flow::data::name%]",
+          "host": "[%key:common::config_flow::data::host%]",
+          "count": "Ping Count"
+        }
```
I think some descriptions here are good, but I'm pretty bad at writing good texts.

jpbede commented:
I think this is ready for review. I hope it is okay to have a config flow only for the binary sensor now.

joostlek commented on homeassistant/components/ping/config_flow.py:
```diff
@@ -0,0 +1,99 @@
+            # import data seems to be a binary_sensor entry
+            self._async_abort_entries_match({CONF_HOST: import_info[CONF_HOST]})
+            return self.async_create_entry(
+                data={"platform_type": "binary_sensor", **import_info}
```
You're importing the data of current entities into `data`, while in the `binary_sensor.py` you get them from `options`. Maybe split up data which can be configured and what not (that's what I've been told is a good practice) between `data` and `options`

jpbede replied:
Hm yes makes sense, then we need to change the function in the parent class `SchemaConfigFlowHandler` first. The function `async_create_entry` does not allow to set `data` and `options` separately:

https://github.com/home-assistant/core/blob/651d4134cf8344171e7350079fb2de1a53db3fb2/homeassistant/helpers/schema_config_entry_flow.py#L339-L348

joostlek commented on homeassistant/components/ping/binary_sensor.py:
```diff
@@ -91,11 +134,17 @@ class PingBinarySensor(RestoreEntity, BinarySensorEntity):
+        self,
+        name: str,
+        ping: PingDataSubProcess | PingDataICMPLib,
+        entry_id: str | None,
```
Why can entry_id be None?

jpbede replied:
Copy&Paste error

joostlek reviewed (changes requested):
Some initial feedback. I haven't seen this way of config flow before so I am not sure if I am the right person to take a look at that

jpbede committed:
Implement review comments

jpbede committed:
Fix comment in config flow

jpbede committed:
Import CONF_PING_COUNT from .const instead redefine it

jpbede committed:
Add test to cover import step

joostlek commented on homeassistant/components/ping/config_flow.py:
```diff
@@ -0,0 +1,99 @@
+
+        if CONF_HOST in import_info:
+            # import data seems to be a binary_sensor entry
+            self._async_abort_entries_match({CONF_HOST: import_info[CONF_HOST]})
```
This method requires you to have `already_configured` in the strings.json

jpbede replied:
Thanks, added.

joostlek reviewed (changes requested):
Last thing from me

jpbede committed:
Add test to cover options flow

gjohansson-ST commented:
We need to migrate the whole thing. Just doing partial bits would be very strange and complicated later on.

jpbede commented:
> We need to migrate the whole thing. Just doing partial bits would be very strange and complicated later on.

Okay, then I'll also migrate the `device_tracker`.

jpbede committed:
Add already_configured abort reason to strings.json

jpbede committed:
Some more ❤️ for the ping integration and also migrate device_tracker

jpbede commented:
> Give some more love to the ping integration. Added configuration flow and migration for binary_sensor and device_tracker platforms. Since both platforms do almost the same thing, I simplified both platforms and used one coordinator.
>
> When migrating, a config entry per host will be created and HA will always set up both platforms. Depending on where the import came from, one of the platforms will be disabled by default.
>
> For newly created entries, the device_tracker platform is disabled by default.

Chose a different approach and gave the integration some more ❤️.

jpbede commented:
This may also fix the issue https://github.com/home-assistant/core/issues/97313, as the ping is run outside the event loop.

jpbede commented:
And https://github.com/home-assistant/core/issues/95047 as well

gjohansson-ST reviewed (changes requested):
This PR is too big.
Please at least remove coordinator implementation as it can be done separately in a PR.
Also this does not comply with https://github.com/home-assistant/architecture/blob/master/adr/0021-YAML-integration-configuration-deprecation-policy.md

jpbede committed:
Remove coordinator from the PR

jpbede committed:
Some cleanup

jpbede committed:
Migrate to asyncio.timeout
